File: examples/node/src/basic/consumer.ts
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { Client, PollingStrategy, Consumer } from 'apache-iggy';
import debug from 'debug';
import { STREAM_ID, TOPIC_ID, PARTITION_ID } from '../utils';

const log = debug('iggy:examples:basic-consumer');


const BATCHES_LIMIT = 5;
const MESSAGES_PER_BATCH = 10;

function parseArgs() {
  const args = process.argv.slice(2);
  const connectionString = args[0] || 'iggy+tcp://iggy:iggy@127.0.0.1:8090';

  if (args.length > 0 && (args[0] === '-h' || args[0] === '--help')) {
    log('Usage: node consumer.js [connection_string]');
    log('Example: node consumer.js iggy+tcp://iggy:iggy@127.0.0.1:8090');
    process.exit(0);
  }

  return { connectionString };
}


async function consumeMessages(client: Client) {
  const interval = 500; // 500 milliseconds
  log(
    'Messages will be consumed from stream: %d, topic: %d, partition: %d with interval %d ms.',
    STREAM_ID,
    TOPIC_ID,
    PARTITION_ID,
    interval
  );

  let offset = 0;
  let consumedBatches = 0;

  while (consumedBatches < BATCHES_LIMIT) {
    try {
      log('Polling for messages...');
      const polledMessages = await client.message.poll({
        streamId: STREAM_ID,
        topicId: TOPIC_ID,
        consumer: Consumer.Single,
        partitionId: PARTITION_ID,
        pollingStrategy: PollingStrategy.Offset(BigInt(offset)),
        count: MESSAGES_PER_BATCH,
        autocommit: false
      });

      if (!polledMessages || polledMessages.messages.length === 0) {
        log('No messages found in current poll - this is expected if the producer had issues sending messages');
        continue;
      }

      offset += polledMessages.messages.length;

      for (const message of polledMessages.messages) {
        const payload = message.payload.toString();
        const { offset, timestamp } = message.headers;

        log('Received message: %s (offset: %d, timestamp: %d)', payload, offset, timestamp);
      }
      log('Consumed %d message(s).', polledMessages.messages.length);
    } catch (error) {
      log('Error consuming messages: %o', error);
    } finally {
      consumedBatches++;
      log('Completed poll attempt %d.', consumedBatches);
      await new Promise(resolve => setTimeout(resolve, interval));
    }
  }

  log('Consumed %d batches of messages, exiting.', consumedBatches);
}

async function main() {
  const args = parseArgs();

  log('Using connection string: %s', args.connectionString);

  // Parse connection string (simplified parsing for this example)
  const url = new URL(args.connectionString.replace('iggy+tcp://', 'http://'));
  const host = url.hostname;
  const port = parseInt(url.port) || 8090;
  const username = url.username || 'iggy';
  const password = url.password || 'iggy';

  const client = new Client({
    transport: 'TCP',
    options: { port, host },
    credentials: { username, password }
  });

  try {
    log('Basic consumer has started, selected transport: TCP');
    // Client connects automatically when first command is called
    log('Connected successfully.');

    // Login will be handled automatically by the client on first command

    await consumeMessages(client);
  } catch (error) {
    log('Error in main: %o', error);
    process.exitCode = 1;
  } finally {
    await client.destroy();
    log('Disconnected from server.');
  }
}


process.on('unhandledRejection', (reason, promise) => {
  log('Unhandled Rejection at: %o, reason: %o', promise, reason);
  process.exitCode = 1;
});

// Prefer async/await style over Promise chaining for consistency
main()

File: core/binary_protocol/src/primitives/consumer.rs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

use crate::WireError;
use crate::WireIdentifier;
use crate::codec::{WireDecode, WireEncode, read_u8};
use bytes::{BufMut, BytesMut};

const KIND_CONSUMER: u8 = 1;
const KIND_CONSUMER_GROUP: u8 = 2;

/// Wire consumer type. Identifies either a single consumer or a consumer group.
///
/// Wire format: `[kind:1][identifier:variable]`
/// - kind=1: consumer
/// - kind=2: consumer group
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct WireConsumer {
    pub kind: u8,
    pub id: WireIdentifier,
}

impl WireConsumer {
    #[must_use]
    pub const fn consumer(id: WireIdentifier) -> Self {
        Self {
            kind: KIND_CONSUMER,
            id,
        }
    }

    #[must_use]
    pub const fn consumer_group(id: WireIdentifier) -> Self {
        Self {
            kind: KIND_CONSUMER_GROUP,
            id,
        }
    }
}

impl WireEncode for WireConsumer {
    fn encoded_size(&self) -> usize {
        1 + self.id.encoded_size()
    }

    fn encode(&self, buf: &mut BytesMut) {
        buf.put_u8(self.kind);
        self.id.encode(buf);
    }
}

impl WireDecode for WireConsumer {
    fn decode(buf: &[u8]) -> Result<(Self, usize), WireError> {
        let kind = read_u8(buf, 0)?;
        if kind != KIND_CONSUMER && kind != KIND_CONSUMER_GROUP {
            return Err(WireError::UnknownDiscriminant {
                type_name: "WireConsumer",
                value: kind,
                offset: 0,
            });
        }
        let (id, id_consumed) = WireIdentifier::decode(&buf[1..])?;
        Ok((Self { kind, id }, 1 + id_consumed))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn roundtrip_consumer() {
        let c = WireConsumer::consumer(WireIdentifier::numeric(42));
        let bytes = c.to_bytes();
        assert_eq!(bytes.len(), 1 + 6);
        let (decoded, consumed) = WireConsumer::decode(&bytes).unwrap();
        assert_eq!(consumed, bytes.len());
        assert_eq!(decoded, c);
    }

    #[test]
    fn roundtrip_consumer_group() {
        let c = WireConsumer::consumer_group(WireIdentifier::numeric(7));
        let bytes = c.to_bytes();
        let (decoded, consumed) = WireConsumer::decode(&bytes).unwrap();
        assert_eq!(consumed, bytes.len());
        assert_eq!(decoded, c);
        assert_eq!(decoded.kind, KIND_CONSUMER_GROUP);
    }

    #[test]
    fn roundtrip_with_string_identifier() {
        let c = WireConsumer::consumer(WireIdentifier::named("my-consumer").unwrap());
        let bytes = c.to_bytes();
        let (decoded, consumed) = WireConsumer::decode(&bytes).unwrap();
        assert_eq!(consumed, bytes.len());
        assert_eq!(decoded, c);
    }

    #[test]
    fn unknown_kind_rejected() {
        let mut bytes = BytesMut::new();
        bytes.put_u8(0xFF);
        WireIdentifier::numeric(1).encode(&mut bytes);
        assert!(WireConsumer::decode(&bytes).is_err());
    }

    #[test]
    fn truncated_buffer() {
        let c = WireConsumer::consumer(WireIdentifier::numeric(1));
        let bytes = c.to_bytes();
        for i in 0..bytes.len() {
            assert!(
                WireConsumer::decode(&bytes[..i]).is_err(),
                "expected error for truncation at byte {i}"
            );
        }
    }
}
